multichannel_timeseries/streaming.ipynb

```diff
@@ -0,0 +1,802 @@
+    "                    'ch': ch,\n",
+    "                })\n",
+    "    \n",
+    "    def get_channel_names(self):\n",
+    "        return self.channel_names\n",
+    "    \n",
+    "    def get_channel_positions(self):\n",
+    "        return self.channel_positions\n",
+    "        \n",
+    "    @property\n",
+    "    def sampling_interval(self):\n",
+    "        return self._sampling_interval\n",
+    "        \n",
+    "    def start(self):\n",
+    "        if not self.streaming:\n",
+    "            self.player.start()\n",
+    "            # Allow time for stream initialization\n",
+    "            time.sleep(0.1)\n",
+    "            \n",
+    "            try:\n",
+    "                self.stream.connect(timeout=5.0)\n",
+    "            except RuntimeError as e:\n",
+    "                print(f\"Failed to connect to LSL stream: {e}\")\n",
+    "                self.player.stop()\n",
+    "                raise\n",
+    "                \n",
+    "            self.stream.pick(self.picks)\n",
+    "            if self.reference:\n",
+    "                self.stream.set_eeg_reference(self.reference)\n",
+    "            self.streaming = True\n",
+    "    \n",
+    "    def stop(self):\n",
+    "        if self.streaming:\n",
+    "            self.stream.disconnect()\n",
+    "            self.player.stop()\n",
+    "            self.streaming = False\n",
+    "    \n",
+    "    def generate_data(self):\n",
+    "        if not self.streaming:\n",
+    "            return pd.DataFrame(columns=['time'] + self.channel_names)\n",
+    "        \n",
+    "        # Collect all new samples since last call\n",
+    "        data, ts = self.stream.get_data(\n",
+    "            self.stream.n_new_samples / self.stream.info[\"sfreq\"],\n",
+    "            picks=self.channel_names\n",
+    "        )\n",
+    "        \n",
+    "        if data.size > 0:\n",
+    "            # Convert timestamps to pandas Timestamps\n",
+    "            ts = pd.to_datetime(ts, unit='s')\n",
+    "            new_data = pd.DataFrame({'time': ts})\n",
+    "            for i, ch in enumerate(self.channel_names):\n",
+    "                new_data[ch] = data[i]\n",
+    "            return new_data\n",
+    "        else:\n",
+    "            return pd.DataFrame(columns=['time'] + self.channel_names)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e8c4104b-0e12-40e3-a647-df4b5a16d930",
+   "metadata": {},
+   "source": [
+    "## Building the Streaming Application\n",
+    "The StreamingApp class below handles the user interface and streaming logic, integrating the data sources with interactive controls. Comments are included for each part of the code that might need additional context."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7845800c-84ba-461f-8f78-5e2bac9b9180",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class StreamingApp(param.Parameterized):\n",
```

Comment:
It's a pretty big class. Removing CPU usage would make it a bit shorter.

Reply:
I'll move streaming to another example